app/src/main/java/com/zeapo/pwdstore/git/sshj/OpenKeychainKeyProvider.kt

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
+ * SPDX-License-Identifier: GPL-3.0-only
+ */
+package com.zeapo.pwdstore.git.sshj
+
+import android.app.PendingIntent
+import android.content.Intent
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.IntentSenderRequest
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.content.edit
+import androidx.preference.PreferenceManager
+import com.github.ajalt.timberkt.d
+import com.zeapo.pwdstore.utils.OPENPGP_PROVIDER
+import com.zeapo.pwdstore.utils.PreferenceKeys
+import java.io.Closeable
+import java.security.PublicKey
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import net.schmizz.sshj.common.Base64
+import net.schmizz.sshj.common.Buffer
+import net.schmizz.sshj.common.DisconnectReason
+import net.schmizz.sshj.common.KeyType
+import net.schmizz.sshj.userauth.UserAuthException
+import net.schmizz.sshj.userauth.keyprovider.KeyProvider
+import org.openintents.ssh.authentication.ISshAuthenticationService
+import org.openintents.ssh.authentication.SshAuthenticationApi
+import org.openintents.ssh.authentication.SshAuthenticationApiError
+import org.openintents.ssh.authentication.SshAuthenticationConnection
+import org.openintents.ssh.authentication.request.KeySelectionRequest
+import org.openintents.ssh.authentication.request.Request
+import org.openintents.ssh.authentication.request.SigningRequest
+import org.openintents.ssh.authentication.request.SshPublicKeyRequest
+import org.openintents.ssh.authentication.response.KeySelectionResponse
+import org.openintents.ssh.authentication.response.Response
+import org.openintents.ssh.authentication.response.SigningResponse
+import org.openintents.ssh.authentication.response.SshPublicKeyResponse
+
+class OpenKeychainKeyProvider(private val activity: AppCompatActivity) : KeyProvider, Closeable {
+
+    private sealed class ApiResponse {
+        data class Success(val response: Response) : ApiResponse()
+        data class GeneralError(val exception: Exception) : ApiResponse()
+        data class NoSuchKey(val exception: Exception) : ApiResponse()
+    }
+
+    private val context = activity.applicationContext
+    private val sshServiceConnection = SshAuthenticationConnection(context, OPENPGP_PROVIDER)
+    private val preferences = PreferenceManager.getDefaultSharedPreferences(context)
+
+    private lateinit var continueAfterUserInteraction: ActivityResultLauncher<IntentSenderRequest>
+    private lateinit var sshServiceApi: SshAuthenticationApi
+
+    private var currentCont: Continuation<Intent>? = null
+    private var keyId
+        get() = preferences.getString(PreferenceKeys.SSH_OPENKEYSTORE_KEYID, null)
```

Comment:
`GitOperation` is accessing `SSH_OPENKEYSTORE_KEYID` from the EncryptedPreferences file, here we're using the standard preferences, which one do we want to stick with?

Reply:
Good catch, I think standard preferences are reasonable here (and backwards compatible).